### PR TITLE
[Example] Sample-based GAT on ogbn-product dataset

### PR DESCRIPTION
**Pull request by classicsong**
## Description
This gives an example of sample-based GAT on ogbn-product dataset.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/ogb/ogbn-products/gat/main.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            if fanout == 0:
+                frontier = dgl.in_subgraph(self.g, seeds)
+            else:
+                frontier = dgl.sampling.sample_neighbors(self.g, seeds, fanout, replace=True)
+            # Then we compact the frontier into a bipartite graph for message passing.
+            block = dgl.to_block(frontier, seeds)
+            # Obtain the seed nodes for next layer.
+            seeds = block.srcdata[dgl.NID]
+
+            blocks.insert(0, block)
```
[the remark this answers is not in the document]
Comment (classicsong) [reply examples/pytorch/ogb/ogbn-products/gat/main.py:41]: Put the block in the block list (blocks are sampled layers)

File: examples/pytorch/ogb/ogbn-products/gat/main.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+#### Entry point
+def run(args, device, data):
+    # Unpack data
+    train_nid, val_nid, test_nid, in_feats, labels, n_classes, g, num_heads = data
+
+    # Create sampler
+    sampler = NeighborSampler(g, [int(fanout) for fanout in args.fan_out.split(',')])
+
+    # Create PyTorch DataLoader for constructing blocks
+    dataloader = DataLoader(
```
[the remark this answers is not in the document]
Comment (classicsong) [reply examples/pytorch/ogb/ogbn-products/gat/main.py:174]: It seems it does not introduce and performance gain.

File: examples/pytorch/ogb/ogbn-products/gat/main.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        batch_size=args.batch_size,
+        collate_fn=sampler.sample_blocks,
+        shuffle=True,
+        drop_last=False,
+        num_workers=args.num_workers)
+
+    # Define model and optimizer
+    model = GAT(in_feats, args.num_hidden, n_classes, args.num_layers, num_heads, F.relu, args.dropout)
+    model = model.to(device)
+    loss_fcn = nn.CrossEntropyLoss()
```
[the remark this answers is not in the document]
Comment (classicsong) [reply examples/pytorch/ogb/ogbn-products/gat/main.py:185]: Emm.. it should be nll_loss

[the remark this answers is not in the document]
Comment (yzh119) [reply examples/pytorch/ogb/ogbn-products/gat/main.py]: Have the results been changed?

**Pull request merged**